New forall & reducer API (#1004)
LLNL/RAJA: docs/sphinx/user_guide/feature/reduction.rst

For more information about available RAJA reduction policies and guidance
on which to use with RAJA execution policies, please see 
:ref:`reducepolicy-label`.

--------------------------------
Experimental Reduction Interface
--------------------------------

> An experimental reduction interface is now available that hopes to improve

Reviewer: Can we add a paragraph to briefly explain how the new reducers work from the user perspective and how maybe how that differs from the old reducers?

Author: Done.

Reviewer 2: @mdavis36 When describe the differences between the new interface and the old one, I think it would be good to add that no reduction policy is required for the new interface.

Author: Done

upon the current reduction model in RAJA. This interface allows ``RAJA::forall``
to take optional "plugin-like" objects to extend the execution behaviour 
of a ``RAJA::forall`` execution context.

RAJA::expt::Reduce
..................
::

>   double rs;
>   double rm;

Reviewer: ```suggestion
  double rs = 0.0;
  double rm = 0.0;
```

Author: Done.

      
  RAJA::forall<EXEC_POL> ( Res, Seg, 
  RAJA::expt::Reduce<RAJA::operators::plus>(&rs),
  RAJA::expt::Reduce<RAJA::operators::minimum>(&rm),
  [=] (int i, double& _rs, double& _rm) {
    _rs += ...
    _rm = RAJA_MIN(..., _rm); 
  }
  );
  
  std::cout << rs ...
  std::cout << rm ...

> 

Reviewer: I want to have the explanation here but I think its more detail than a user needs to think about when they're using the reducers. It would also help to have an example as the "local variable" mentioned here is implicit in the example. I like going into detail about how the reference stuff works, but I think that we should have a more detailed example like an example implementation with all the pieces if we're going to talk about the pieces.

Author: I'm not quite sure what you're asking for here

Author: @MrBurmark With the most recent changes to the documentation, do you still want this comment to be addressed. I believe the documentation is in a detailed enough state to demonstrate how the reducers work from a users perspective. Is there something specific you want changed / added.

> * ``RAJA::expt::Reduce`` takes a target variable to write the final result to 

Reviewer: I think we should be clear that Reduce combines its value with the existing value instead of just overwriting it.
```suggestion
* ``RAJA::expt::Reduce`` takes a target variable to add the final result to 
```

Reviewer 3: Hmm, it might be worth having that be an option, extra arg to reduce for initial value that defaults to the original?

Author: Done.

  (``rs``, ``rm``).

> * It passes a corresponding argument to the RAJA lambda to be used as the 

Reviewer: What is "It" here, this sentence is unclear.

Author: Done.

  local instance of the target(``_rs``, ``_rm``).

> * The local variable is initialized with the "identity" of the reduction 

Reviewer: I always think this means that it initializes rs and rm, could it be more clear which variable is set here?

Author: Done.

For more information about available RAJA reduction policies and guidance
on which to use with RAJA execution policies, please see 
:ref:`reducepolicy-label`.

--------------------------------
Experimental Reduction Interface
--------------------------------

> An experimental reduction interface is now available that hopes to improve

Reviewer 2: ```suggestion
An experimental reduction interface is now available that offers several usability and performance advantages over
```

Author: Done

> upon the current reduction model in RAJA. This interface allows ``RAJA::forall``

Reviewer 2: ```suggestion
the current reduction model in RAJA. This interface allows ``RAJA::forall``
```

Author: Done

to take optional "plugin-like" objects to extend the execution behaviour 
of a ``RAJA::forall`` execution context.

> The new interface passes ``RAJA::expt::Reduce<OP_TYPE>`` objects as direct

Reviewer 2: ```suggestion
The new interface passes ``RAJA::expt::Reduce<OP_TYPE>`` objects as function arguments
```

Author: Done

> parameters to ``RAJA::forall`` and provides users with thread-local variables

Reviewer 2: ```suggestion
to ``RAJA::forall`` and provides users with thread-local variables
```

Author: Done

> of the operating data type to be manipulated indside the lambda. This differs 

Reviewer 2: ```suggestion
of the reduction data type to be updated inside the lambda. This differs 
```

Author: Done

from the current reduction model in which ``RAJA::ReduceOP<REDUCE_POL, T>``

> objects are captured by the RAJA lambda and manipulated directly.

Reviewer 2: ```suggestion
objects are captured by the user-supplied kernel body lambda expression.
```

Author: Done

RAJA::expt::Reduce
..................
::

  double rs = 0.0;
  double rm = 1e100;
      
  RAJA::forall<EXEC_POL> ( Res, Seg, 
  RAJA::expt::Reduce<RAJA::operators::plus>(&rs),
  RAJA::expt::Reduce<RAJA::operators::minimum>(&rm),
  [=] (int i, double& _rs, double& _rm) {

>     _rs += ...

Reviewer 2: It would be much clearer and help users if you replaced the ellipses ('...') with actual arguments. For example, here add double* a = ...; before the kernel and use a[i] in the reduction updates in the kernel.

Also, please add a note that 'Res' is a resource object that must be compatible with the exec pol, and 'Seg' is the iteration space object.

Author: Done

    _rm = RAJA_MIN(..., _rm); 
  }
  );
  
  std::cout << rs ...
  std::cout << rm ...

* ``RAJA::expt::Reduce`` takes a target variable to write the final result to 
  (``rs``, ``rm``). The reduction operation will include the initial value of
  the given target variable.
* ``RAJA::forall`` requires a corresponding lambda argument to
  ``RAJA::expt::Reduce`` be passed (``_rs``, ``_rm``). These arguments can be
  used as the local instance of the target for the reduction operation.
* The local variables (``_rs, ``_rm``) are initialized with the "identity" of
  the reduction operation to be performed.
* A reduction is performed implicitly by the ``RAJA::forall`` across thread 
  copies of the local variables.
* Finally, the reduction operation is performed against the original value of 
  the target and the result of the ``RAJA::forall`` reduction.

> * The final value can be returned simply be referencing the target variable.

Reviewer 2: ```suggestion
* The final reduction value is accessed by referencing the target variable passed to the ``RAJA::forall`` method.
```

Author: Done

RAJA::expt::ValLoc
..................

> RAJA supports ``Loc`` reductions. With this new interface ``Loc`` reductions 

Reviewer 2: ```suggestion
RAJA supports ``Loc`` reductions, which provide the ability to get a kernel/loop index at which the final reduction value was found. With this new interface ``Loc`` reductions 
```

Reviewer 2: @mdavis36 does this change the semantics of 'loc' reductions? For example, If I initialize the value of a min reduction to -1 and I perform the reduction over an array of positive values, I will get the index of the min value in the array via getLoc(). Is that the same behavior as the current model? I can't remember.

Author: If you set the initial value of the target ValLoc to -1 and perform a min reduction on an array of positive integer you will get -1. This is the same behavior as with our current reduction model. You must initialize the value to above the highest value in the array if you wish to find the min loc in the array. Tested with this:
```
  RAJA::TypedRangeSegment<int> testrange(0, 10);
  std::array<int, 10> test = {43,7,6,1,9,12,14,46,78,3};

  std::cout << "\n Running RAJA OLD LOC omp reductions...\n";
  RAJA::ReduceMinLoc< RAJA::omp_reduce, int > vminloc(-1, -1);

  RAJA::forall<RAJA::omp_parallel_for_exec>( testrange,
    [=](RAJA::Index_type i) {

    vminloc.minloc( test[i], i );

  });
  std::cout << "Old min : " << vminloc.get() << "\n";
  std::cout << "Old loc : " << vminloc.getLoc() << "\n";

  std::cout << "\n Running RAJA NEW LOC omp reductions...\n";
  VALLOC_INT new_minloc(-1, -1);

  RAJA::forall<RAJA::omp_parallel_for_exec>(testrange,
    RAJA::expt::Reduce<RAJA::operators::minimum>(&new_minloc),
    [=](int i, VALLOC_INT& _new_minloc) {
      _new_minloc.min(test[i], i);
      //_new_minloc = RAJA_MIN(VALLOC_INT(test[i], i), _new_minloc);
    }
  );

  std::cout << "New min : " << new_minloc.getVal() << "\n";
  std::cout << "New loc : " << new_minloc.getLoc() << "\n";
```
Output:
```
 Running RAJA OLD LOC omp reductions...
Old min : -1
Old loc : -1

 Running RAJA NEW LOC omp reductions...
New min : -1
New loc : -1
```

can be performed using ``ValLoc<T>`` types. Since they are strongly typed they

> provide min() and max() operations. Users must also use getVal() and getLoc to

Reviewer 2: ```suggestion
provide min() and max() operations. Users must also use ``getVal()`` and ``getLoc()`` methods to
```

> return results.

Reviewer 2: ```suggestion
access the reduction results.
```

Author: Done

::

  using VL_INT = RAJA::expt::ValLoc<int>;
  VL_INT rm_loc;
      
  RAJA::forall<EXEC_POL> ( Res, Seg, 
  RAJA::expt::Reduce<RAJA::operators::minimum>(&rm_loc),
  [=] (int i, VL_INT& _rm_loc) {

>     _rm_loc = RAJA_MIN(..., _rm_loc);  

Reviewer 2: Same comment about replacing ellipses with arguments

Author: Done.

  }
  );

  std::cout << rm_loc.getVal() ...
  std::cout << rm_loc.getLoc() ...

Lambda Arguments
................

> This interface takes advantage of C++ parameter packs to allow users to define

Reviewer 2: ```suggestion
This interface takes advantage of C++ parameter packs to allow users to pass
```

Author: Done

> any number of ``expt::Reduce`` objects in their RAJA::forall calls.

Reviewer 2: ```suggestion
any number of ``RAJA::expt::Reduce`` or ``RAJA::expt::ValLoc`` objects to their RAJA::forall calls.
```

Author: Done

::

  using VL_INT = RAJA::expt::ValLoc<int>;
  VL_INT rm_loc;
  double rs;
  double rm;
        
  RAJA::forall<EXEC_POL> ( Res, Seg, 
    RAJA::expt::Reduce<RAJA::operators::plus>(&rs),        // --> 1 double added
    RAJA::expt::Reduce<RAJA::operators::minimum>(&rm),     // --> 1 double added
    RAJA::expt::Reduce<RAJA::operators::minimum>(&rm_loc), // --> 1 VL_INT added
    RAJA::expt::KernelName("MyFirstRAJAKernel"),           // --> NO args added
    [=] (int i, double& _rs, double& _rm, VL_INT& _rm_loc) {

>       _rs += ...

Reviewer 2: Same comment about replacing ellipses with arguments

Author: Done.

--------------------------------

An experimental reduction interface is now available that offers several 
usability and performance advantages over the current reduction model in RAJA. 
The new interface allows ``RAJA::forall`` to take optional "plugin-like" 
objects to extend the execution behavior of a ``RAJA::forall`` execution 
context.

The new interface passes ``RAJA::expt::Reduce<OP_TYPE>`` objects as function 
arguments to ``RAJA::forall`` and provides users with thread-local variables
of the reduction data type to be updated inside the lambda. This differs 
from the current reduction model in which ``RAJA::ReduceOP<REDUCE_POL, T>``
objects are captured by the user-supplied kernel body lambda expression.

RAJA::expt::Reduce
..................
::

  double* a = ...;

  double rs = 0.0;
  double rm = 1e100;
      
  RAJA::forall<EXEC_POL> ( Res, Seg, 
  RAJA::expt::Reduce<RAJA::operators::plus>(&rs),
  RAJA::expt::Reduce<RAJA::operators::minimum>(&rm),
  [=] (int i, double& _rs, double& _rm) {
    _rs += a[i];
    _rm = RAJA_MIN(a[i], _rm); 
  }
  );
  
  std::cout << rs ...
  std::cout << rm ...

* Each ``RAJA::expt::Reduce`` argument to ``RAJA::forall`` is templated on

>   a reduction operator, and takes a reference to a target variable to write 

Reviewer: ```suggestion
  a reduction operator, and takes a pointer to a target variable to write 
```

Reviewer 2: Good catch. My bad.

  the final reduction result to, ``&rs`` and ``&rm`` in the example code
  above. The reduction operation will include the initial value of
  the given target variable.

> * The kernel body lambda expression passed to ``RAJA::forall`` requires an 

Reviewer: ```suggestion
* The kernel body lambda expression passed to ``RAJA::forall`` must take an 
```

> * The kernel body lambda expression passed to ``RAJA::forall`` requires an 
>   argument corresponding to each ``RAJA::expt::Reduce`` argument, ``_rs`` and 
>   ``_rm`` in the example code. These arguments are used as a local instance of 
>   of each target of a reduction operation. It is important to note that the 
>   the lambda expression reduction arguments must follow the kernel iteration
>   variable, ``i`` in this case, and appear in the same order as the 
>   corresponding ``RAJA::expt::Reduce`` arguments to ``RAJA::forall``. They
>   also must have the same types as the values passed to the 
>   ``RAJA::expt::Reduce`` arguments. 

Reviewer: ```suggestion
* The kernel body lambda expression passed to ``RAJA::forall`` must have a 
  parameter corresponding to each ``RAJA::expt::Reduce`` argument, ``_rs`` and 
  ``_rm`` in the example code. These parameters refer to a local target of each
  reduction operation. It is important to note that the parameters follow the
  kernel iteration variable, ``i`` in this case, and appear in the same order as the 
  corresponding ``RAJA::expt::Reduce`` arguments to ``RAJA::forall``. Their
  types must be references to the types passed to the  ``RAJA::expt::Reduce``
  arguments. 
```

  RAJA::expt::Reduce<RAJA::operators::minimum>(&rm),
  [=] (int i, double& _rs, double& _rm) {
    _rs += a[i];
    _rm = RAJA_MIN(a[i], _rm); 
  }
  );
  
  std::cout << rs ...
  std::cout << rm ...

* Each ``RAJA::expt::Reduce`` argument to ``RAJA::forall`` is templated on
  a reduction operator, and takes a reference to a target variable to write 
  the final reduction result to, ``&rs`` and ``&rm`` in the example code
  above. The reduction operation will include the initial value of
  the given target variable.
* The kernel body lambda expression passed to ``RAJA::forall`` requires an 
  argument corresponding to each ``RAJA::expt::Reduce`` argument, ``_rs`` and 
  ``_rm`` in the example code. These arguments are used as a local instance of 
  of each target of a reduction operation. It is important to note that the 
  the lambda expression reduction arguments must follow the kernel iteration
  variable, ``i`` in this case, and appear in the same order as the 
  corresponding ``RAJA::expt::Reduce`` arguments to ``RAJA::forall``. They
  also must have the same types as the values passed to the 
  ``RAJA::expt::Reduce`` arguments. 

> * The local variables ``_rs`` and ``_rm`` are initialized with the *identity* of
>   the reduction operation to be performed.

Reviewer: ```suggestion
* The local variables referred to by ``_rs`` and ``_rm`` are initialized with the
  *identity* of the reduction operation to be performed.
```

> * A reduction is performed implicitly by the ``RAJA::forall`` execution
>   across thread copies of the local variables.

Reviewer: ```suggestion
* The local variables are updated in the user supplied lambda as it is called.
* The local variables are reduced to a single value, combining their values across all
  threads participating in the ``RAJA::forall`` execution.
  .
```

> * Finally, the reduction operation is performed against the original value of 
>   the target and the result of the ``RAJA::forall`` reduction.

Reviewer: ```suggestion
* Finally, the target variable is updated with the result of the ``RAJA::forall`` reduction
  by performing the reduction operation to combine the existing value of the target 
  variable and the result of the ``RAJA::forall`` reduction.
```

> * The final reduction value is accessed by referencing the target variable 
>   passed to ``RAJA::expt::Reduce`` in the ``RAJA::forall`` method.

Reviewer: ```suggestion
* The target variable passed to the ``RAJA::expt::Reduce`` argument
  of ``RAJA::forall`` then has the final reduction value.
```

Author: I'm not sure about this one, but I took the rest

  [=] (int i, double& _rs, double& _rm) {
    _rs += a[i];
    _rm = RAJA_MIN(a[i], _rm); 
  }
  );
  
  std::cout << rs ...
  std::cout << rm ...

* Each ``RAJA::expt::Reduce`` argument to ``RAJA::forall`` is templated on
  a reduction operator, and takes a pointer to a target variable to write 
  the final reduction result to, ``&rs`` and ``&rm`` in the example code
  above. The reduction operation will include the existing value of
  the given target variable.
* The kernel body lambda expression passed to ``RAJA::forall`` must have a 
  parameter corresponding to each ``RAJA::expt::Reduce`` argument, ``_rs`` and 
  ``_rm`` in the example code. These parameters refer to a local target for each
  reduction operation. It is important to note that the parameters follow the
  kernel iteration variable, ``i`` in this case, and appear in the same order as the 
  corresponding ``RAJA::expt::Reduce`` arguments to ``RAJA::forall``. The
  parameters' types must be references to the types used in the
  ``RAJA::expt::Reduce`` arguments.
* The local variables referred to by ``_rs`` and ``_rm`` are initialized with the
  *identity* of the reduction operation to be performed.

> * The local variables are updated in the user supplied lambda as it is called.

Reviewer: ```suggestion
* The local variables are updated in the user supplied lambda.
```

* The local variables are reduced to a single value, combining their values across all
  threads participating in the ``RAJA::forall`` execution.

>   .

Reviewer: ```suggestion
```

  variable and the result of the ``RAJA::forall`` reduction.
* The final reduction value is accessed by referencing the target variable 
  passed to ``RAJA::expt::Reduce`` in the ``RAJA::forall`` method.

.. note:: In the above example ``Res`` is a resource object that must be 
          compatible with the ``EXEC_POL``. ``Seg`` is the iteration space
          object for ``RAJA::forall``.

.. important:: The order and types of the local reduction variables in the
               kernel body lambda expression must match exactly with the
               corresponding ``RAJA::expt::Reduce`` arguments to the
               ``RAJA::forall`` to ensure that the correct result is obtained.

RAJA::expt::ValLoc
..................

As with the current RAJA reduction interface, the new interface supports *loc* 
reductions, which provide the ability to get a kernel/loop index at which the 
final reduction value was found. With this new interface, *loc* reductions 
are performed using ``ValLoc<T>`` types. Since they are strongly typed, they
provide ``min()`` and ``max()`` operations that are equivalent to using 
``RAJA_MIN()`` or ``RAJA_MAX`` macros as demonstrated in the code example below.

> Users must use ``getVal()`` and ``getLoc()`` methods to access the reduction 

Reviewer: ```suggestion
Users must use the ``getVal()`` and ``getLoc()`` methods to access the reduction 
```

  }
  );

  std::cout << rm_loc.getVal() ...
  std::cout << rm_loc.getLoc() ...

Lambda Arguments
................

This interface takes advantage of C++ parameter packs to allow users to pass
any number of ``RAJA::expt::Reduce`` objects to the ``RAJA::forall`` method::

  double* a = ...;

  using VL_DOUBLE = RAJA::expt::ValLoc<double>;
  VL_DOUBLE rm_loc;
  double rs;
  double rm;
        
  RAJA::forall<EXEC_POL> ( Res, Seg, 
    RAJA::expt::Reduce<RAJA::operators::plus>(&rs),        // --> 1 double added
    RAJA::expt::Reduce<RAJA::operators::minimum>(&rm),     // --> 1 double added
    RAJA::expt::Reduce<RAJA::operators::minimum>(&rm_loc), // --> 1 VL_DOUBLE added
    RAJA::expt::KernelName("MyFirstRAJAKernel"),           // --> NO args added
    [=] (int i, double& _rs, double& _rm, VL_DOUBLE& _rm_loc) {
      _rs += a[i];
      _rm = RAJA_MIN(a[i], _rm); 
      _rm_loc.min(VL_DOUBLE(a[i], i));
    }
  );

  std::cout << rs ...
  std::cout << rm ...
  std::cout << rm_loc.getVal() ...
  std::cout << rm_loc.getLoc() ...

Again, the lambda expression arguments are passed in the same order as
the ``RAJA::expt::Reduce`` arguments to ``RAJA::forall``. Both the types and 

> order of arguments must match to get correct results and to compile 

Reviewer: ```suggestion
order of the parameters must match to get correct results and to compile 
```

successfully. Otherwise, a static assertion will be triggered::

  LAMBDA Not invocable w/ EXPECTED_ARGS.

> .. note:: This static assert is only enabled when passing a generic lambda.

Reviewer: I'm not sure generic lambda is the phrasing you want to use here. Generic is used with lambda to mean using auto at the param type.

Author: I used generic as in "not extended". I believe the CUDA documentation refers to lambdas like this as well, so that was the convention I followed.

Reviewer: How about non-device?

Author: non device lambdas are not considered extended lambdas by CUDA

Reviewer: Wait, does generic lambda mean a host device lambda in cuda?

Reviewer 2: I don't think it does. I think 'generic lambda' refers to the 'auto' param type(s) as you described. I believe that the CUDA 'extended lambda' concept refers to being able to decorate a lambda with __host__ __device__ and use them in either host or device execution context.

          This check will not happen when passing extended-lambdas (i.e. DEVICE
          tagged lambdas) or other functor like objects.

.. note:: The experimental ``RAJA::forall`` interface is more flexible than the
          current implementation, other optional arguments besides
          ``RAJA::expt::Reduce`` can be passed to a ``RAJA::forall`` to extend
          its behavior. In the above example we demonstrate using 
          ``RAJA::expt::KernelName``, which wraps a ``RAJA::forall`` executing 

>           under a ``HIP`` or ``CUDA`` policy in a named region. Use of the 
>           ``RAJA::expt::KernelName`` parameter does not require any aditional
>           arguments to the lambda expression.

Reviewer: ```suggestion
          under a ``HIP`` or ``CUDA`` policy in a named region. Use of 
          ``RAJA::expt::KernelName`` does not require an additional
          parameter in the lambda expression.
```
